Combat tracker

Ampersandrew opened the pull request:
I have most of this written. I think it's ready for incorporation. Still needs other bits done, but wht's here is mostly working.

Ampersandrew committed:
Make all existing files pass linting checks

Also add in some ignopre patterns

Ampersandrew committed:
Add in the guts of the combat tracker

This needs a lot of testing and front end work.

It has a deck class that manages a deck of cards for one of the
two factions involved in a combat.

It has a hand class that manages a set of cards belonging to one
participant in a fight.

The cards module has meta information about a standard deck
including a mapping from index to symbol (two characters) to
name (a string).

The two main classes are Deadlands Classic overrides of combat
and combatant. Combat manages the two decks and each combatant
manages a hand of cards.

Ampersandrew committed:
Tidying up combat

Tweaks to the round management.

There is now a payuse at the beginning of a round to roll quickness
and draw cards.

Ampersandrew committed:
Restructure the module folder

Move files into documents and helper folders.
Update all the import statement.

Ampersandrew committed:
Move to deadlands fix bugs

deadlands-classic didn't work everywhere, sometimes it was
interpreted as deadlands - classic. So change to deadlands.

Fix lots of bugs.

Ampersandrew committed:
Change game system directory, add in sidebar

1. Rename the game mode to deadlands-classic.
1. a) Update all strings and paths to the new name.

2. Restructure the scripts into modules since we are using
   the new syntax.

3. Subclass CombatTracker sidebar class to change location of html

4. Add in html for the combat tracker
4. a) add labels for the combat tracker

5. in DeadlandsCombat, make resetRoundData and sortCombatants private.

6. Add offerEndTurn to DeadlandsCombat. When all normal cards have been
   used, there may be jokers or sleeved cards left, the combat must be
   manually moved to the stage where new cards are drawn.

7. Make contents, hasBlackJoker, hasNormal, hasRedJoker, hasSleeved, and
   roundStarted properties of combatant (delegates to embeded hand object).

7. a) make toggleBlackJoker, toggleRedJoker, toggleSleeved methods of combatant
   (delegates to embeded hand object).

8. Improve the fromObject method of Hand. When an object goes the the become
   a flag, retrieve a flag process, it loses its class. This lets us construct
   a Hand object from the generic Javascript object retrieved from the flag.

Ampersandrew committed:
Tweaks to the html

trying to get the combat tracker html to do sensible things.

Ampersandrew committed:
Add card images

Add in a set of card images in svg format.

Dulux-Oz committed:
20230906b - MJB - Added Card Images

Added 2nd set of 50 card images

Dulux-Oz committed:
20230906c - MJB - Added Card Images

Added last set of card images

Dulux-Oz committed:
20230906a - MJB - Added Card Images

Added 1st set of 50 card images

Ampersandrew committed:
Add the path to the icon file

Is now available as aCard[inx].icon where inx is an integer
in the range 0 .. 53

Ampersandrew committed:
Fix bugs add cards and images to handlebar data

Ampersandrew committed:
Tidy up access to composed class

use the cards accessor on combatant to get the imgaes instead of
going direct to the ahnd.

Ampersandrew committed:
Fix bug using wrong decks

The combatants were using the wrong decks.

Ampersandrew committed:
Make Card images configurable

Add a game setting to choose between the Deadlands cards and
the plainer "Modern" cards.

Generate the icon fields of the cards based on the choice.

Note: need to restart the application after changing this. It is a local
setting and each player can suit themselves, the setting is saved on
their machine, not in the global database. This also means you don't
need elevated permissions to set it for yourself.

Ampersandrew committed:
Actually call the icon generation

Call the operation that generates the card image paths

Ampersandrew committed:
Fix broken deepclone

Foundry lies about its deepclone method, it claims to cope with arrays.

It leaves arrays linked to the original instead of separating them into
separately modifiable things.

Ampersandrew committed:
Move Turn events

Copy this from the base class so we can tweak it.

Ampersandrew committed:
Fix the combat history

Store more of the combat history and allow it to persis
between sessions.

Fix some bugs where arrays were pushed to arrays instead of being
expanded and then pushed.

Ampersandrew committed:
Sort out the cards display and undoing turns

Each combatant now has a display of its active cards.

Have improved the turn undo (previous turn) function. Seems to be
working.

Ampersandrew committed:
Various bug fixes in dlc-combat

Add in a hasPreviousTurns getter.

Make sortCombatants a method. Instead of everywhere just doing
this.turns.sort directly,

Fix previous turn, so it sorts the combatants.

Fix the sorting of combatants in setupTurns. Was causing strange issues
when the round was started and a sleeved card was toggled in and out of
use.

Fix #sortCombatants, when sleeved cards and joker are toggled into use,
initiative becomes a number which is not a card. Explicitly checking for
initiative being a card was forcing them to the wrong end of the sort
order.

Ampersandrew committed:
Tidyup deck

Remove methods which turn the deck into a string and back
into a deck. These were written to usewhen storing the deck
in a flag. They are not necessary as full objects can be stored
in a document's flags.

Move some methods for better organisation.

Ampersandrew committed:
Improve storage and retrieval of round state

Make the state of the round into one item in the flags, the
individual items of this state are never stored and retrieved
separately.

Move the retrieval of round data into prepare turns.

Ampersandrew committed:
Bugfix to combatant

Fix options that were causig the order of combatants to change while
drawing cards between rounds.

Ampersandrew committed:
Combatant bugfix for using sleeved

Need to check that override is not -1. Otherwise it is true when
hand is completely empty and held and override are both -1.

Ampersandrew committed:
Add in socket stuff

Improved the html of the combat tracker

Add in a socket implementation so that players can affect the combat
without needing extra permissions.

Ampersandrew committed:
Improve shuffling behaviour

Combatant
=========
Change cleanupRound to return the cards cleaned up instead of discarding
them back to their deck of origin. Also, don't save parent state to
flags when called, parent will store state if necessary.

Combat
======
Stop storing the decks' shuffle flags to the Combat state flags. These
are really only used in resetRoundData and there seems to be a bug in
Foundry that was clearing them.

Merge resetRoundData and reconcileDeck.

Tidy up reapDiscards to not continually write combat's flags, once per
combatant.

Deck
====
Remove the setting of the shuffle flag while drawing. This is now only
ever set by reconcile, and is then immediately used by the combat in
resetRoundData.

Change discard to use a filter to eliminate duplicates from discards. It
no longer returns the number of cards discarded as nothing was using
it.

Ampersandrew committed:
Make it loadable

modify the system.json so I can load it off github

Ampersandrew committed:
Refactoring

Deck
----
* Remove dead code.

Hand
----
* Rename live -> cards
         spent -> discards
* Remove unused purge method

Combat
------
* Rename deck allies -> ally
* Change cleanall to use new combatant method that returns cards instead
  of discarding them back to their deck.
* Change vamoose to be a method on combat, not combatant
* Move draw, sleeveHighest, toggleHostility, toggleSleeved,
  toggleRedJoker
  and toggleBlackJoker from combatant, since they save combat state and
  should be combat methods.

Combatant
---------
* Removed all parent updates from methods. All the methods were called
  from the combat-tracker and should have been combat methods.

socket-functions
----------------
Changed a lot of methods to call combat methods and pass the combatant
id
instead of looking up and calling combatant methods. All these methods
were modifying combat state.

Ampersandrew committed:
Implement delete and undelete

Fixed the addition of Jokers to the hands when drawing cards.

Added the ability to delete and undelete cards. It is not possible
to delete or undelete a joker.

Ampersandrew committed:
Tidy up flag access

Change the method of determining whether a flag is present on the
document and retrieving its value.

Ampersandrew committed:
Make npcs start hostile

Add an on create method that makes npcs initially hostile when
added to the combat.

Ampersandrew committed:
Fix setFlag bug

Calling setFlags multiple times was writing one flag and then extracting
all the other flags from the database and updating the combat. This was
overwritting the data before it could be read.

Changed to use an update that writes all of the flags simultaneously.

Improved the turn event handling.

Ampersandrew committed:
Fix some combat tracker bugs

Now properly handles end of turn when there are slleved cards or
jokers.

Some tweaks to the character sheet.

Add a configuration file, really belongs in character sheet, but
has a banner I've added.

Some language additions also really for character sheet, but doing
no harm in the combat tracker.

Change the icon for draw cards to the one I added for it, stop using
the d20.

TheSpyder commented on module/helpers/hand.mjs:
```diff
@@ -0,0 +1,421 @@
+  }
+
+  static fromObject(proto) {
+    const isHostile =
+      typeof proto !== 'undefined' &&
+      typeof proto.isHostile !== 'undefined' &&
+      typeof proto.isHostile === 'boolean' &&
+      proto.isHostile;
```
There are probably a few areas we could make fancy improvements, but this is an obvious one 😂 
```suggestion
    const isHostile = Boolean(proto?.isHostile);
```

Ampersandrew replied:
Yes, if you look at the commit called tidy up flag access you can see me tidy up lots of instances of that kind of thing.

TheSpyder reviewed (dismissed):
Some of it is a bit big to review, but I think it looks reasonable enough that we should run it on Matt's server and see what we think

Ampersandrew committed:
Update module/helpers/hand.mjs

Tidy up the nonsense. Deal with undefined values in an idiomatic way.

Co-authored-by: Andrew Herron <thespyder@programmer.net>

haxxxton commented:
@Ampersandrew, im logged in as the GM, and am getting errors when trying to draw cards for my actors.

```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
    at DeadlandsCombatTracker._onCombatantControl
```

Appears the `game['deadlands-classic'].socket` is undefined when trying to `drawCard` (line 236 of dlc-combat-tracker.mjs).

Am i doing something wrong?

Ampersandrew commented:
I might not have committed all the files needed. Let me have a look.


-- 
  Andrew Wilson
  ***@***.***


On Sun, 8 Oct 2023, at 6:03 AM, haxxxton wrote:
> 
> 
> @Ampersandrew <https://github.com/Ampersandrew>, im logged in as the GM, and am getting errors when trying to draw cards for my actors.
> 
> `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
>     at DeadlandsCombatTracker._onCombatantControl
`
> Appears the `game['deadlands-classic'].socket` is undefined when trying to `drawCard` (line 236 of dlc-combat-tracker.mjs).
> 
> Am i doing something wrong?
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Dulux-Oz/DeadlandsClassic/pull/7#issuecomment-1751921875>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABDARTTMST7YVOKB4UFRMUDX6IXYZAVCNFSM6AAAAAA5MQHAFWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONJRHEZDCOBXGU>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

Ampersandrew commented:
Ah no. There 's a module you have to install to get the sockets working. Sorry I forgot to mention it.

Install socketlib, that'll sort it out. I was going to reimplement this, but after I dissected it, it was just
going to waste time, so I decided to use the library.

-- 
  Andrew Wilson
  ***@***.***


On Sun, 8 Oct 2023, at 6:03 AM, haxxxton wrote:
> 
> 
> @Ampersandrew <https://github.com/Ampersandrew>, im logged in as the GM, and am getting errors when trying to draw cards for my actors.
> 
> `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
>     at DeadlandsCombatTracker._onCombatantControl
`
> Appears the `game['deadlands-classic'].socket` is undefined when trying to `drawCard` (line 236 of dlc-combat-tracker.mjs).
> 
> Am i doing something wrong?
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Dulux-Oz/DeadlandsClassic/pull/7#issuecomment-1751921875>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABDARTTMST7YVOKB4UFRMUDX6IXYZAVCNFSM6AAAAAA5MQHAFWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONJRHEZDCOBXGU>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

Ampersandrew commented:
Background. using cards in any way affects the combat. Only GMs can modify the
combat. Players can modify their own participant in the combat (the combatant).
I wrote it and checked it out as I was going, I had most of it working using the GM.

Then I tried it logged in as a player and couldn't draw or use cards, because doing
that affected the combatant order which is a combat function.

I made every thing use a socket so that even the GMs client asks "A connected
GM" to make the change. This will still be done on their system, but will go through
a socket.

There are some changes that only a GM should legitimately make, the buttons
that invoke those functions are not shown to connected players, only to connected
GMs.

-- 
  Andrew Wilson
  ***@***.***


On Sun, 8 Oct 2023, at 3:58 PM, Andrew Wilson wrote:
> Ah no. There 's a module you have to install to get the sockets working. Sorry I forgot to mention it.
> 
> Install socketlib, that'll sort it out. I was going to reimplement this, but after I dissected it, it was just
> going to waste time, so I decided to use the library.
> 
> -- 
>   Andrew Wilson
>   ***@***.***
> 
> 
> On Sun, 8 Oct 2023, at 6:03 AM, haxxxton wrote:
>> 
>> 
>> @Ampersandrew <https://github.com/Ampersandrew>, im logged in as the GM, and am getting errors when trying to draw cards for my actors.
>> 
>> `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
>>     at DeadlandsCombatTracker._onCombatantControl
`
>> Appears the `game['deadlands-classic'].socket` is undefined when trying to `drawCard` (line 236 of dlc-combat-tracker.mjs).
>> 
>> Am i doing something wrong?
>> 
>> 
>> 
>> —
>> Reply to this email directly, view it on GitHub <https://github.com/Dulux-Oz/DeadlandsClassic/pull/7#issuecomment-1751921875>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABDARTTMST7YVOKB4UFRMUDX6IXYZAVCNFSM6AAAAAA5MQHAFWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONJRHEZDCOBXGU>.
>> You are receiving this because you were mentioned.Message ID: ***@***.***>
>> 
>> 
>

haxxxton commented:
@Ampersandrew is socketlib an npm package? how is that being included in the game system? (is there a build step that isnt in the package.json?), or should i be copying over the node_modules folder to the system folder in fftv?

Ampersandrew commented:
> @Ampersandrew is socketlib an npm package? how is that being included in the game system? (is there a build step that isnt in the package.json?), or should i be copying over the node_modules folder to the system folder in fftv?

It's a foundry extension module, you install it with the Foundry application. When you fire Foundry up it gives you a screen with three options. Game worlds, Game systems and Add-on Modules. It probably should be in the package.json somewhere that the game system depends on the module. I'll have a look.

Ampersandrew committed:
Clone the socketlib stuff into the system

To avoid having external dependencies the socketlib library
has been cloned into our system. The MIT lisence allows for
this.

Ampersandrew commented:
I have cloned the socket stuff into our system, and brought it up to our coding standards. This is allowed by its original MIT license. This means that we no longer need to install socketlib for our system to run.

Ampersandrew committed:
Bugfix Sleeving not undoable

Sleeving a card was not storing enough state to allow it to be
undone.

While fixing that noticed that nextTurn, Vamoose and sleeveCard
had a lot of common code, refactored the common stuff. As part of
this changed the sleeveHightest socket operation to not take
a combatantId as only the acting combatant can sleeve a card.

Also noticed that the now extracted common saveTurn function
had a lot in common with previousTurn. Extracted the common
bits into another function.

Removed some redundant stuff in the socket implementation.

Ampersandrew committed:
Tweaks

Changed an instance of an inconsistent icon.

Fixed a bit of logic.

TheSpyder reviewed (approved):
If there was some pre-requisite steps for installation that would've been fine, rather than vendoring the code into our project, but we can clean that up later

Ampersandrew commented:
I had originally started cleaning it, I was interested to see how it works, and then decided I should just use it. That's what I originally put up.  Since I didn't know how to tell Foundry that it needs the library and it doesn't fall over in a nice way, it wasn't obvious to James why it wasn't running.

I mentioned on the #system-developer discord channel that I was planning have the system use an external module. No one there thought it was a good idea to have the system dependent on someone else's code. I asked how to tell Foundry that the system needs a module to run, they universally thought that was a bad idea.

Based on that, I cleaned it up and added it to our system.

TheSpyder commented:
Ah that makes sense, thank you!

TheSpyder commented on system.json:
```diff
@@ -48,10 +48,8 @@
-  "url": "https://github.com/Dulux-Oz/DeadlandsClassic",
-  "manifest": "https://raw.githubusercontent.com/Dulux-Oz/DeadlandsClassic/master/system.json",
-  "download": "https://github.com/Dulux-Oz/DeadlandsClassic/releases/download/0.1/deadlands_classic.zip",
+  "url": "https://github.com/Ampersandrew/deadlands-classic",
+  "manifest": "https://raw.githubusercontent.com/Ampersandrew/deadlands-classic/CombatTracker/system.json",
+  "download": "https://github.com/Ampersandrew/deadlands-classic/releases/download/0.1/deadlands_classic.zip",
```
Now that I know what I'm looking at, these values are pointing to your github repo not this repo 🤔

Ampersandrew replied:
I was trying to get the thing to install from github. That was part of me trying to get it to work. It only does that when your module is registered with Foundry. I forgot this was in here. It should be backed out.

TheSpyder replied:
After I wrote that I realised I should be using a local checkout of the code, not telling it to install from the `system.json` link. So there's no rush, bigger problems to fix than this right now 😅
